iwyu.cc

```diff
@@ -2043,21 +2043,87 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     if (CanIgnoreCurrentASTNode())
       return true;
 
-    // We assume that all type traits with >= 2 arguments, except '__is_same',
-    // require full type information even for pointer types. For example,
-    // this is the case for `__is_convertible_to` trait.
-    if (expr == nullptr || expr->getNumArgs() < 2 ||
-        expr->getTrait() == TypeTrait::BTT_IsSame) {
+    if (expr == nullptr || expr->getNumArgs() < 2)
       return true;
-    }
 
-    for (const TypeSourceInfo* arg : expr->getArgs()) {
-      QualType qual_type = arg->getType();
-      const Type* type = qual_type.getTypePtr();
-      ReportTypeUse(CurrentLoc(), type, DerefKind::RemoveRefsAndPtr);
-    }
+    const Type* lhs_type = expr->getArg(0)->getType().getTypePtr();
+    const Type* rhs_type = nullptr;
+    if (expr->getNumArgs() == 2)
+      rhs_type = expr->getArg(1)->getType().getTypePtr();
+
+    // Most of the binary type traits require the types to be complete types or
+    // arrays of unknown bound for compilable code. It means that compilers
+    // accept pointers or references to fwd-declared types. However, in certain
+    // cases it may cause a UB (C++20 [meta.rqmts]p5) because the full
+    // pointed-to type is needed to determine the trait value correctly, so it's
+    // better for IWYU to require the full type info in such situations. It is
+    // important not to suggest a full type when it is redundant for computing
+    // the correct value.
+    switch (expr->getTrait()) {
+      case TypeTrait::BTT_IsSame:
+        return true;
+      case TypeTrait::BTT_IsAssignable:
+      case TypeTrait::BTT_IsNothrowAssignable:
+        // TODO(bolshakov): require complete argument types here when
+        // CanForwardDeclareType stops doing it.
+        if (RemoveReference(lhs_type)->isRecordType()) {
```

Comment:
I think you want to check for `TagType` here to also cover unions. Or `isRecordType() || isUnionType()`.

Reply:
Wait, what? `RecordType` also covers unions? I was under the impression that unions were the difference between `Tag` and `Record` types. But it doesn't look like it, so I guess this works fine.

Reply:
Yes, `RecordType` covers classes, structs, and unions, and `TagType` also covers enums.